src/workerd/api/sockets.c++

```diff
@@ -47,52 +47,68 @@ jsg::Ref<Socket> connectImpl(
   return connectImplNoOutputLock(js, kj::mv(actualFetcher), kj::mv(address));
 }
 
-InitData initialiseSocket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise) {
+InitData initialiseSocket(
+    jsg::Lock& js, kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise,
+    std::function<void()> onClose) {
   auto& context = IoContext::current();
 
   // Initialise the readable/writable streams with a custom AsyncIoStream that waits for the
   // completion of `connectionPromise` before performing reads/writes.
-  auto stream = kj::refcounted<PipelinedAsyncIoStream>(kj::mv(connectionPromise));
+  auto stream = kj::refcounted<PipelinedAsyncIoStream>(kj::mv(connectionPromise), onClose);
   auto sysStreams = newSystemMultiStream(kj::addRef(*stream), StreamEncoding::IDENTITY, context);
+  auto readable = jsg::alloc<ReadableStream>(context, kj::mv(sysStreams.readable));
+  auto writable = jsg::alloc<WritableStream>(context, kj::mv(sysStreams.writable));
+
+  auto closeFulfiller = IoContext::current().addObject(
+      kj::heap<kj::PromiseFulfillerPair<void>>(kj::newPromiseAndFulfiller<void>()));
+  auto closeFulfillerFork = closeFulfiller->promise.fork();
+
+  // Attach a callback to close the readable/writable streams when the socket is closed (either
+  // explicitly or implicitly).
+  context.awaitIo(js, closeFulfillerFork.addBranch(), [readable = readable.addRef(),
+      writable = writable.addRef()](jsg::Lock& js) mutable {
+    readable->cancel(js, nullptr);
+    writable->abort(js, nullptr);
+  });
```

Comment:
Nothing is holding onto the JS promise returned by `awaitIo` here, and more importantly nothing is handling if it happens to reject (which is entirely possible).

There's also a challenge here using `readable->cancel(...)` and `writable->abort(...)` like this. Each of those return their own promises that will reject if the `readable` / `writable` are locked to `Reader` / `Writer`. You'll likely want to use the `abort`/`cancel` on the underlying controllers instead, and you'll definitely want to add error handling here. If you don't care if this JS promise rejects then you can explicitly mark is handled (there's an API for that in `jsg::Promise`)

Reply:
Right. IIRC from what we discussed we want the `closed` promise to be resolved with these errors.

Can you elaborate on what you mean by the "underlying controllers"?

Reply:
See the `getController()` methods on both `ReadableStream` and `WritableStream`

Reply:
I added error handling to this and ended up only closing the streams when the socket is explicitly closed. This means calling `close()` on the socket will close the underlying streams, but when the readable stream reads 0 bytes the `closed` promise will resolve with the underlying streams remaining unclosed. I'm not sure these are semantics we want but hopefully this is good enough for now and we can adjust these later.